### DISCUSSION TITLE
Confused about the best way to provide an OpenAI api key

### DISCUSSION THREAD

**Discussion by Martinsos** [karthink/gptel#518] [General]
Hi there, thanks for this awesome package!

I am just setting it up, and I wanted to start with using GPT as a model. When looking at docs in the readme, for setting up ChatGPT, they say to procure the OpenAI API key, which makes sense, but then the only other thing is "Optional: set gptel-api-key" and two bullet points below it. This confused me, due to the "Optional" part -> I assumed this means that it can work without this, probably pulling it in directly from the `OPENAI_API_KEY` env var. But it seems like that is not happening. It did ask me interactively for the GPT API Key, but that lasted only for the current emacs session. So what is the recommended way to do it? Thanks!

**Comment by tillydray** [comment]
One of the best ways to handle secrets in emacs, in general, is [`authinfo.gpg`](https://www.emacswiki.org/emacs/GnusAuthinfo). You create `~/.authinfo.gpg` on your computer, then add a line for each secret you want. It's outlined in the readme [here](https://github.com/karthink/gptel/blob/55b9f618b9c4abb34207e82cf61df7f3209ea0d1/README.org#chatgpt)

**Comment by karthink** [comment]
You can put

```elisp
(setq gptel-api-key "sk-blahblabblah...")
```
in your gptel configuration.  But then the key will be visible in your config, which is unsafe.

Instead, you can do
```elisp
(setq gptel-api-key (getenv "OPENAI_API_KEY"))
```
and make sure that this environment variable is available to Emacs.

If your API keys are stored in a password manager of some kind, you can do
```elisp
(setq gptel-api-key (lambda () (shell-command-to-string "my-password-get api.openai.com")))
```
where `my-password-get` is a shell command that retrieves your passwords.

But the best way to do it would be to use the `.authinfo.gpg` file as explained by @tillydray.

**Comment by Martinsos** [comment]
Thanks @karthink ! That makes sense, thanks. Would it make sense to update README to make this more obvious? What especially confused me is the "optional" word, since I didn't get what is the non-optional part of it, is that a mistake in README?

**Comment by amake** [comment]
I would like to point out that auth-source supports the macOS keychain, so you can store your key there if you are a Mac user:

1. Create key at https://platform.openai.com/api-keys
2. Open the Keychain Access app
3. Create new password item
4. Make sure it's recognized as an Internet password by entering
   `https://api.openai.com` for the Keychain Item Name
5. Use `apikey` as Account Name
6. Enter key as password and save
7. Test by executing `security find-internet-password -g -s api.openai.com -a apikey` in a shell

**Comment by vedang** [comment]
Along similar lines, I opened a PR with an example custom function for `gptel-api-key` to fetch this from ENV vars: https://github.com/karthink/gptel/pull/1170
